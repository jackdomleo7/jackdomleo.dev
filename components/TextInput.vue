<template>
  <div class="text-input">
    <label :for="id" class="text-input__label">
      {{ label }}
      <template v-if="required">
        <span v-if="!hideRequiredAsterisk" class="text-input__asterisk" aria-hidden="true">
          *
        </span> <span class="sr-only">
          (required)
        </span>
      </template>
    </label>
    <p class="text-input__help">
      {{ helpText }}
    </p>
    <span class="text-input__entry" :class="{ 'text-input__entry--error': errorMsg }">
      <input
        :id="id"
        :value="value"
        :name="name"
        :type="type"
        :inputmode="inputmode"
        :autocomplete="autocomplete"
        :placeholder="placeholder"
        :required="required"
        :aria-invalid="errorMsg ? 'true' : 'false'"
        :aria-describedby="errorMsg && `${id}--error`"
        @change="$emit('change', $event.target.value)"
        @input="$emit('input', $event.target.value)"
        @blur="$emit('blur', $event.target.value)"
      >
    </span>
    <p :id="`${id}--error`" class="text-input__error" :role="errorMsg && 'alert'" :aria-hidden="errorMsg ? 'false' : 'true'">
      {{ errorMsg }}
    </p>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { stringHasNoWhitespace } from '@/helpers/propValidators'

export default Vue.extend({
  name: 'TextInput',
  props: {
    label: {
      type: String,
      required: true
    },
    id: {
      type: String,
      required: true,
      validator (value: string): boolean {
        return stringHasNoWhitespace(value)
      }
    },
    value: {
      type: String,
      default: undefined
    },
    type: {
      type: String,
      default: 'text',
      validator: (value: string): boolean => {
        return ['text', 'email'].includes(value) && stringHasNoWhitespace(value)
      }
    },
    inputmode: {
      type: String,
      default: undefined,
      validator: (value: string): boolean => {
        return ['text', 'email'].includes(value) && stringHasNoWhitespace(value)
      }
    },
    required: {
      type: Boolean,
      default: false
    },
    hideRequiredAsterisk: {
      type: Boolean,
      default: false
    },
    name: {
      type: String,
      default: undefined,
      validator (value: string): boolean {
        return stringHasNoWhitespace(value)
      }
    },
    placeholder: {
      type: String,
      default: undefined
    },
    autocomplete: {
      type: String,
      default: undefined
    },
    helpText: {
      type: String,
      default: undefined
    },
    errorMsg: {
      type: String,
      default: undefined
    }
  }
})
</script>

<style lang="scss" scoped>
.text-input {
  display: flex;
  flex-direction: column;
  gap: 0.125rem;

  &__label {
    color: var(--colour-text-primary);
    text-transform: uppercase;
    font-size: 0.875rem;
    font-weight: 700;
    user-select: none;
  }

  &__asterisk {
    color: var(--colour-primary);
    font-size: 700;
  }

  &__help {
    margin: 0;
    color: var(--colour-text-secondary);
    font-size: 0.75rem;
  }

  &__entry {
    border: 1px solid var(--colour-text-secondary);
    background-color: var(--colour-foreground-1);
    border-radius: 0.1875rem;
    height: 3rem;

    &--error {
      border-color: var(--colour-error);
    }

    input {
      color: var(--colour-text-primary);
      font-size: 1rem;
      height: 100%;
      width: 100%;
      padding-inline: 0.5rem;
      background-color: transparent;
      border: none;

      &::placeholder {
        color: #b6b5b4;
      }
    }
  }

  &__error {
    font-size: 0.75rem;
    margin-bottom: 0;
    margin-top: 0.1875rem;
    color: var(--colour-error);
  }
}
</style>
