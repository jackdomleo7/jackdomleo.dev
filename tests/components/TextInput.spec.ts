import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import TextInput from '@/components/TextInput.vue'

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('TextInput', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Emitters
   */

  it('emits the change event on the input', async () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    await wrapper.find('input').trigger('change')

    expect(wrapper.emitted('change')).toBeTruthy()
  })

  it('emits the input event on the input', async () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    await wrapper.find('input').trigger('input')

    expect(wrapper.emitted('input')).toBeTruthy()
  })

  it('emits the blur event on the input', async () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter'
      }
    })

    await wrapper.find('input').trigger('blur')

    expect(wrapper.emitted('blur')).toBeTruthy()
  })

  /**
   * v-model
   */

  it('should set its input element\'s value to string "Jack" when v-model equals string "Jack"', () => {
    const wrapper = mount({
      data: () => ({ userName: 'Jack' }),
      components: { TextInput },
      template: '<TextInput v-model="userName" label="Name" id="user-name" />'
    })

    expect((wrapper.find('input').element as HTMLInputElement).value).toBe('Jack')
  })

  it('should alter its input element\'s value to "Domleo" when the v-model value is changed from "Jack" to "Domleo"', async () => {
    const wrapper = mount({
      components: { TextInput },
      data: () => ({ name: 'Jack' }),
      template: '<TextInput v-model="name" label="Name" id="user-name" />'
    })

    const input = wrapper.find('input')

    expect((input.element as HTMLInputElement).value).toBe('Jack')

    await wrapper.setData({ name: 'Domleo' })

    expect((input.element as HTMLInputElement).value).toBe('Domleo')
  })

  /**
   * Props
   */

  it('sets the label to "Name" when set', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name'
      }
    })

    expect(wrapper.find('label').text()).toBe('Name')
  })

  it('sets the id when set', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name'
      }
    })

    expect(wrapper.find('label').attributes('for')).toBe('name')
    expect(wrapper.find('input').attributes('id')).toBe('name')
  })

  it('sets the type to "text" when type is not set', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name'
      }
    })

    expect(wrapper.find('input').attributes('type')).toBe('text')
  })

  it('sets the type to "text" when type is set to "text"', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        type: 'text'
      }
    })

    expect(wrapper.find('input').attributes('type')).toBe('text')
  })

  it('sets the type to "email" when type is set to "email"', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        type: 'email'
      }
    })

    expect(wrapper.find('input').attributes('type')).toBe('email')
  })

  it('sets the inputmode to "text" when inputmode is set to "text"', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        inputmode: 'text'
      }
    })

    expect(wrapper.find('input').attributes('inputmode')).toBe('text')
  })

  it('sets the inputmode to "email" when inputmode is set to "email"', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        inputmode: 'email'
      }
    })

    expect(wrapper.find('input').attributes('inputmode')).toBe('email')
  })

  it('renders a required input', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        required: true
      }
    })

    expect(wrapper.find('label').text()).toContain('*')
    expect(wrapper.find('label').text()).toContain('(required)')
    expect(wrapper.find('input').attributes('required')).toBe('required')
  })
  

  it('renders a required input with a hidden asterisk', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        required: true,
        hideRequiredAsterisk: true
      }
    })

    expect(wrapper.find('label').text()).not.toContain('*')
    expect(wrapper.find('label').text()).toContain('(required)')
    expect(wrapper.find('input').attributes('required')).toBe('required')
  })

  it('sets the name attribute on the input', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Name',
        id: 'name',
        name: 'first-name'
      }
    })

    expect(wrapper.find('input').attributes('name')).toBe('first-name')
  })

  it('sets the placeholder attribute on the input', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Email',
        id: 'email',
        placeholder: 'test@example.com'
      }
    })

    expect(wrapper.find('input').attributes('placeholder')).toBe('test@example.com')
  })

  it('sets the autocomplete attribute on the input', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Email',
        id: 'email',
        autocomplete: 'email'
      }
    })

    expect(wrapper.find('input').attributes('autocomplete')).toBe('email')
  })

  it('renders help text when set', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter',
        type: 'email',
        helpText: 'No spam. Unsubscribe anytime.'
      }
    })

    expect(wrapper.find('p.text-input__help').text()).toBe('No spam. Unsubscribe anytime.')
  })

  it('renders an error state and an error message', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter',
        type: 'email',
        errorMsg: 'Please provide a valid email'
      }
    })

    expect(wrapper.find('p#newsletter--error').text()).toBe('Please provide a valid email')
    expect(wrapper.find('p#newsletter--error').attributes('role')).toBe('alert')
    expect(wrapper.find('p#newsletter--error').attributes('aria-hidden')).toBe('false')
    expect(wrapper.find('span.text-input__entry').classes()).toContain('text-input__entry--error')
    expect(wrapper.find('input').attributes('aria-invalid')).toBe('true')
    expect(wrapper.find('input').attributes('aria-describedby')).toBe('newsletter--error')
  })

  it('renders a normal state when no error message is set', () => {
    const wrapper = shallowMount(TextInput, {
      propsData: {
        label: 'Newsletter',
        id: 'newsletter',
        type: 'email'
      }
    })

    expect(wrapper.find('p#newsletter--error').text()).toBe('')
    expect(wrapper.find('p#newsletter--error').attributes('role')).toBeUndefined()
    expect(wrapper.find('p#newsletter--error').attributes('aria-hidden')).toBe('true')
    expect(wrapper.find('span.text-input__entry').classes()).not.toContain('text-input__entry--error')
    expect(wrapper.find('input').attributes('aria-invalid')).toBe('false')
    expect(wrapper.find('input').attributes('aria-describedby')).toBeUndefined()
  })
})
