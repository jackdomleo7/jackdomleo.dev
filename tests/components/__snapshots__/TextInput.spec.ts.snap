// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput matches snapshot when mounted 1`] = `
<div
  class="text-input"
>
  <label
    class="text-input__label"
    for="newsletter"
  >
    
    Newsletter
    
    <!---->
  </label>
   
  <p
    class="text-input__help"
  >
    
    
  
  </p>
   
  <span
    class="text-input__entry"
  >
    <input
      aria-invalid="false"
      id="newsletter"
      type="text"
    />
  </span>
   
  <p
    aria-hidden="true"
    class="text-input__error"
    id="newsletter--error"
  >
    
    
  
  </p>
</div>
`;

exports[`TextInput matches snapshot when shallow mounted 1`] = `
<div
  class="text-input"
>
  <label
    class="text-input__label"
    for="newsletter"
  >
    
    Newsletter
    
    <!---->
  </label>
   
  <p
    class="text-input__help"
  >
    
    
  
  </p>
   
  <span
    class="text-input__entry"
  >
    <input
      aria-invalid="false"
      id="newsletter"
      type="text"
    />
  </span>
   
  <p
    aria-hidden="true"
    class="text-input__error"
    id="newsletter--error"
  >
    
    
  
  </p>
</div>
`;
